rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3597,6 +3793,43 @@ static rmw_ret_t get_topic_name(dds_entity_t endpoint_handle, std::string & name
   } while (true);
 }
 
+static rmw_ret_t check_for_service_reader_writer(const CddsCS & client, bool * is_available)
+{
+  std::vector<dds_instance_handle_t> rds, wrs;
+  if (get_matched_endpoints(client.pub->enth, dds_get_matched_subscriptions, rds) < 0 ||
+    get_matched_endpoints(client.sub->enth, dds_get_matched_publications, wrs) < 0)
+  {
+    RMW_SET_ERROR_MSG("rmw_service_server_is_available: failed to get reader/writer matches");
+    return RMW_RET_ERROR;
+  }
+  // first extract all service ids from matched readers
+  std::set<std::string> needles;
+  for (auto rdih : rds) {
+    auto rd = get_matched_subscription_data(client.pub->enth, rdih);
+    std::string serviceid;
+    if (rd.get() && get_user_data_key(rd->qos, "serviceid", serviceid)) {
+      needles.insert(serviceid);
+    }
+  }
+  if (needles.size() == 0) {
+    return RMW_RET_OK;
+  }
+  // then scan the writers to see if there is at least one with a service id in the set
+  for (auto wrih : wrs) {
+    auto wr = get_matched_publication_data(client.sub->enth, wrih);
+    std::string serviceid;
+    if (wr.get() &&
+      get_user_data_key(
+        wr->qos, "serviceid",
+        serviceid) && needles.find(serviceid) != needles.end())
+    {
+      *is_available = true;
```

Comment:
@eboasson considering asserting that `is_available != nullptr && !*is_available`, which seems to be the implicit assumption in this function.